### PR TITLE
.Net: OpenAPI customization

### PR DESCRIPTION
**Pull request by SergeyMenshykh**
### Motivation and Context
Today, OpenAPI functionality in SK supports a limited, predefined list of content types: text/plain, application/json, etc. Moreover, the way these content types are read from the "wire" might not be the most efficient for certain scenarios. For example, SK OpenAPI functionality reads JSON as a string and images as a byte array. While this may be sufficient for most cases, there could be situations where this approach is not optimal, and reading both JSON and image content as streams would be preferable.

### Description
This PR adds functionality allowing to supply a HTTP content reader to customize HTTP content reading. The reader accepts HTTP request and HTTP response as parameters, so it can access all properties of both to decide on the best way to read content and support content types not supported by default. For example, this custom reader can read JSON as a stream and any content as stream if the `x-stream` heder is provided in HTTP request:
```csharp
private static async Task<object?> ReadHttpResponseContentAsync(HttpResponseContentReaderContext context, CancellationToken cancellationToken)
{
    // Read JSON content as a stream rather than as a string, which is the default behavior
    if (context.Response.Content.Headers.ContentType?.MediaType == "application/json")
    {
        return await context.Response.Content.ReadAsStreamAsync(cancellationToken);
    }

    // Read any content as a stream if the request contains the "x-stream" header.
    if (context.Request.Headers.Contains("x-stream"))
    {
        return await context.Response.Content.ReadAsStreamAsync(cancellationToken);
    }

    // Return null to indicate that any other HTTP content not handled above should be read by the default reader.
    return null;
}

// Register the custom HTTP content reader
var executionParameters = new OpenApiFunctionExecutionParameters() { HttpResponseContentReader = ReadHttpResponseContentAsync };

// Create OpenAPI plugin
var plugin = await OpenApiKernelPluginFactory.CreateFromOpenApiAsync("RepairService", "Resources/Plugins/RepairServicePlugin/repair-service.json", executionParameters);

```

Closes: https://github.com/microsoft/semantic-kernel/issues/8070

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/src/Functions/Functions.OpenApi/RestApiOperationRunner.cs [hunk 1]
```diff
@@ -262,32 +295,12 @@ private static async Task<RestApiOperationResponse> SerializeResponseContentAsyn
 
         var mediaType = contentType?.MediaType ?? throw new KernelException("No media type available.");
```
Comment (RogerBarreto) [inline dotnet/src/Functions/Functions.OpenApi/RestApiOperationRunner.cs:296]: ContentType for text and image will be `text/???`, `image/???` right?

File: dotnet/src/Functions/Functions.OpenApi/RestApiOperationRunner.cs [hunk 2]
```diff
@@ -464,5 +477,60 @@ private Uri BuildsOperationUrl(RestApiOperation operation, IDictionary<string, o
...
+        if (this._httpResponseContentReader is not null)
+        {
+            content = await this._httpResponseContentReader.Invoke(new(requestMessage, responseMessage), cancellationToken).ConfigureAwait(false);
+        }
+
+        // If no custom reader is provided or the custom reader did not return any content, read the content using the default readers.
+        if (content is null)
+        {
+            // Obtain the content reader by media type (e.g., text/plain, application/json, image/jpg)
+            if (!s_contentReaderByContentType.TryGetValue(mediaType, out var reader))
```
Comment (RogerBarreto) [inline dotnet/src/Functions/Functions.OpenApi/RestApiOperationRunner.cs:502]: I don't think you are going to find the keys for `text` and `image` without ignoring the `subtype`

```suggestion
            if (!s_contentReaderByContentType.TryGetValue(mediaType, out var reader))
```

File: dotnet/samples/Concepts/Plugins/OpenApiPlugin_CustomHttpContentReader.cs [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+namespace Plugins;
+
+/// <summary>
+/// Sample shows how to register a custom HTTP content reader for an Open API plugin.
+/// </summary>
+public sealed class CustomHttpContentReaderForOpenApiPlugin(ITestOutputHelper output) : BaseTest(output)
+{
+    [Fact]
+    public async Task ShowReadingJsonAsStreamAsync()
```
Comment (RogerBarreto) [inline dotnet/samples/Concepts/Plugins/OpenApiPlugin_CustomHttpContentReader.cs:16]: Suggest adding the tests for the different supported content types `text/*`, `image/*`, `application/xml` if there aren't any.

File: dotnet/src/Functions/Functions.OpenApi/RestApiOperationRunner.cs (hunk 2 above)
Comment (SergeyMenshykh) [reply dotnet/src/Functions/Functions.OpenApi/RestApiOperationRunner.cs:502]: Right, this line looks for a content reader by the full media type, such as `text/jpeg`. If no exact match is found, the code a few lines below, tries to resolve a reader by the primary content type, such as `text`.

File: dotnet/src/Functions/Functions.OpenApi/RestApiOperationRunner.cs [hunk 4]
```diff
@@ -45,14 +46,17 @@ internal sealed class RestApiOperationRunner
...
+    /// TODO: Pass cancelation tokes to the content readers.
+    /// </remarks>
+    private static readonly Dictionary<string, HttpResponseContentReader> s_contentReaderByContentType = new()
     {
-        { "image", async (content) => await content.ReadAsByteArrayAndTranslateExceptionAsync().ConfigureAwait(false) },
-        { "text", async (content) => await content.ReadAsStringWithExceptionMappingAsync().ConfigureAwait(false) },
-        { "application/json", async (content) => await content.ReadAsStringWithExceptionMappingAsync().ConfigureAwait(false)},
-        { "application/xml", async (content) => await content.ReadAsStringWithExceptionMappingAsync().ConfigureAwait(false)}
+        { "image", async (context, _) => await context.Response.Content.ReadAsByteArrayAndTranslateExceptionAsync().ConfigureAwait(false) },
+        { "text", async (context, _) => await context.Response.Content.ReadAsStringWithExceptionMappingAsync().ConfigureAwait(false) },
```
[the remark this answers is not in the document]
Comment (SergeyMenshykh) [reply dotnet/src/Functions/Functions.OpenApi/RestApiOperationRunner.cs:57]: For this simple scenario, no advanced syntax is required, and a basic exact string match has been working well so far.

File: dotnet/samples/Concepts/Plugins/OpenApiPlugin_CustomHttpContentReader.cs (hunk 3 above)
Comment (SergeyMenshykh) [reply dotnet/samples/Concepts/Plugins/OpenApiPlugin_CustomHttpContentReader.cs:16]: This functionality with the readers (formerly called serializers) has been in SK for a while and is already covered with a sufficient amount of unit tests.
- [ItShouldReadContentAsStringSuccessfullyAsync](https://github.com/microsoft/semantic-kernel/blob/6f3642e2dfd47bd702d07dbb3d45eaff14e801fd/dotnet/src/Functions/Functions.UnitTests/OpenApi/RestApiOperationRunnerTests.cs#L950)
- [ItShouldReadContentAsBytesSuccessfullyAsync](https://github.com/microsoft/semantic-kernel/blob/6f3642e2dfd47bd702d07dbb3d45eaff14e801fd/dotnet/src/Functions/Functions.UnitTests/OpenApi/RestApiOperationRunnerTests.cs#L991)

**Review by markwallace-microsoft — APPROVED** [review]

**Review by RogerBarreto — APPROVED** [review]

**Pull request merged**